### PR TITLE
Enable OAuth login via Auth0

### PR DESCRIPTION
**Pull request by amcclain**
+ Integrate with Auth0 as an OAuth provider to enable end-user login with selected social identity providers.
+ Init new OauthService.js on client `preAuthInit()` to authenticate with and coordinate calls to Auth0 using their SPA API.
+ Decode and validate ID tokens (JWTs) sent to server to set/create server-side user after successful OAuth.
+ Email notification to configured addresses on new user sign-in.
+ Auto-grant APP_READER role to all users. HOIST_ADMIN remains assigned via soft-config.
+ Remove all Bootstrap users other than local admin user, with username and credentials sourced (as before) from instanceConfig.

### CODE REVIEW

File: grails-app/services/io/xh/toolbox/security/Auth0Service.groovy [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                return new JwtValidationResult(token: token, exception: e)
+            }
+        }
+    }
+
+
+    //------------------------
+    // Implementation
+    //------------------------
+    private JSONClient _jsonClient
```
Comment (lbwexler) [inline grails-app/services/io/xh/toolbox/security/Auth0Service.groovy:68]: This looks like it is obsolete

File: grails-app/services/io/xh/toolbox/user/UserService.groovy [hunk 2]
```diff
@@ -1,20 +1,83 @@
...
+
+    private void notifyOnUserCreated(User newUser) {
+        def to = emailService.parseMailConfig('newUserNotificationRecipients')
+        if (to) {
+            emailService.sendEmail(
+                to: to,
+                subject: "New Toolbox user created: ${newUser.username}",
+                html: """
+                    A new user has signed into Toolbox (${Utils.appEnvironment.toString()}): <br>
+                    Email: ${newUser.email}<br>
```
Comment (lbwexler) [inline grails-app/services/io/xh/toolbox/user/UserService.groovy:75]: nice

File: grails-app/services/io/xh/toolbox/security/Auth0Service.groovy [hunk 3]
```diff
@@ -18,6 +18,9 @@ import static io.xh.hoist.util.Utils.withNewSession
 @Slf4j
 class Auth0Service extends BaseService {
 
+
+    static clearCachesConfigs = ['auth0Domain', 'auth0ClientId', 'auth0Jwks']
+
```
Comment (amcclain) [inline grails-app/services/io/xh/toolbox/security/Auth0Service.groovy:23]: Thanks - I forgot about this feature!

File: grails-app/services/io/xh/toolbox/security/Auth0Service.groovy [hunk 4]
```diff
@@ -65,12 +68,6 @@ class Auth0Service extends BaseService {
     //------------------------
     // Implementation
     //------------------------
-    private JSONClient _jsonClient
-    JSONClient getJsonClient() {
-        if (!_jsonClient) _jsonClient = new JSONClient()
-        return _jsonClient
-    }
-
```
Comment (amcclain) [inline grails-app/services/io/xh/toolbox/security/Auth0Service.groovy]: And thanks for mopping this up

**Pull request merged**